File: scripts/add-harvard-course.ts
import { createClient } from '@supabase/supabase-js';
import { readFileSync } from 'fs';
import { v4 as uuidv4 } from 'uuid';

// Load environment variables
import { config } from 'dotenv';
config();

const supabaseUrl = process.env.VITE_SUPABASE_URL || process.env.SUPABASE_URL;
const supabaseKey = process.env.VITE_SUPABASE_PUBLISHABLE_KEY || process.env.VITE_SUPABASE_ANON_KEY || process.env.SUPABASE_ANON_KEY;

if (!supabaseUrl || !supabaseKey) {
  console.error('Missing Supabase credentials. Please set VITE_SUPABASE_URL and VITE_SUPABASE_PUBLISHABLE_KEY environment variables.');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);

interface Course {
  title: string;
  slug: string;
  description: string;
  short_description: string;
  category: string;
  difficulty: string;
  price_gel: number;
  is_active: boolean;
  sort_order: number;
}

interface Section {
  title: string;
  sort_order: number;
  lectures: Lecture[];
}

interface Lecture {
  title: string;
  description: string;
  assignments?: Assignment[];
}

interface Assignment {
  title: string;
  description: string;
}

interface CourseData {
  course: Course;
  sections: Section[];
}

async function loadCourseData(): Promise<CourseData> {
  const data = readFileSync('./harvard_python_course.json', 'utf-8');
  return JSON.parse(data);
}

async function insertCourse(course: Course): Promise<string> {
  const now = new Date().toISOString();
  
  // Check if course already exists
  const { data: existing } = await supabase
    .from('video_courses')
    .select('id')
    .eq('slug', course.slug)
    .single();
  
  let courseId: string;
  
  if (existing) {
    courseId = existing.id;
    console.log(`Course already exists with ID: ${courseId}`);
    
    // Update existing course
    await supabase
      .from('video_courses')
      .update({
        title: course.title,
        description: course.description,
        short_description: course.short_description,
        category: course.category,
        difficulty: course.difficulty,
        price_gel: course.price_gel,
        is_active: course.is_active,
        sort_order: course.sort_order,
        updated_at: now
      })
      .eq('id', courseId);
  } else {
    courseId = uuidv4();
    
    const { error } = await supabase
      .from('video_courses')
      .insert({
        id: courseId,
        title: course.title,
        slug: course.slug,
        description: course.description,
        short_description: course.short_description,
        category: course.category,
        difficulty: course.difficulty,
        price_gel: course.price_gel,
        is_active: course.is_active,
        sort_order: course.sort_order,
        created_at: now,
        updated_at: now
      });
    
    if (error) {
      throw new Error(`Failed to insert course: ${error.message}`);
    }
  }
  
  return courseId;
}

async function insertSections(courseId: string, sections: Section[]): Promise<Map<string, string>> {
  const sectionIds = new Map<string, string>();
  
  for (const section of sections) {
    const sectionId = uuidv4();
    const now = new Date().toISOString();
    
    const { error } = await supabase
      .from('video_course_sections')
      .insert({
        id: sectionId,
        course_id: courseId,
        title: section.title,
        sort_order: section.sort_order,
        created_at: now
      });
    
    if (error) {
      console.error(`Failed to insert section "${section.title}": ${error.message}`);
    } else {
      sectionIds.set(section.title, sectionId);
    }
  }
  
  return sectionIds;
}

async function insertLecturesAndAssignments(
  courseId: string,
  sectionIds: Map<string, string>,
  sections: Section[]
): Promise<void> {
  let lectureCount = 0;
  let assignmentCount = 0;
  
  for (const section of sections) {
    const sectionId = sectionIds.get(section.title);
    if (!sectionId) {
      console.error(`Section ID not found for: ${section.title}`);
      continue;
    }
    
    for (const lecture of section.lectures) {
      const lectureId = uuidv4();
      const now = new Date().toISOString();
      
      // Insert lecture
      const { error: lectureError } = await supabase
        .from('video_lectures')
        .insert({
          id: lectureId,
          section_id: sectionId,
          course_id: courseId,
          title: lecture.title,
          description: lecture.description,
          video_url: null,
          video_storage_path: null,
          duration_seconds: 0,
          sort_order: lectureCount,
          is_free_preview: false,
          created_at: now
        });
      
      if (lectureError) {
        console.error(`Failed to insert lecture "${lecture.title}": ${lectureError.message}`);
      } else {
        lectureCount++;
        
        // Insert assignments
        if (lecture.assignments) {
          for (const assignment of lecture.assignments) {
            const assignmentId = uuidv4();
            
            const { error: assignmentError } = await supabase
              .from('video_assignments')
              .insert({
                id: assignmentId,
                lecture_id: lectureId,
                course_id: courseId,
                title: assignment.title,
                description: assignment.description,
                sort_order: assignmentCount,
                created_at: now
              });
            
            if (assignmentError) {
              console.error(`Failed to insert assignment "${assignment.title}": ${assignmentError.message}`);
            } else {
              assignmentCount++;
            }
          }
        }
      }
    }
  }
  
  console.log(`Inserted ${lectureCount} lectures and ${assignmentCount} assignments`);
}

async function main() {
  try {
    console.log('Loading course data...');
    const courseData = await loadCourseData();
    
    console.log('Inserting course...');
    const courseId = await insertCourse(courseData.course);
    console.log(`Course ID: ${courseId}`);
    
    console.log('Inserting sections...');
    const sectionIds = await insertSections(courseId, courseData.sections);
    console.log(`Inserted ${sectionIds.size} sections`);
    
    console.log('Inserting lectures and assignments...');
    await insertLecturesAndAssignments(courseId, sectionIds, courseData.sections);
    
    console.log('\n✅ Course successfully added to database!');
    console.log(`Course: ${courseData.course.title}`);
    console.log(`Slug: ${courseData.course.slug}`);
    console.log(`Total sections: ${courseData.sections.length}`);
    
    const totalLectures = courseData.sections.reduce((sum, s) => sum + s.lectures.length, 0);
    const totalAssignments = courseData.sections.reduce(
      (sum, s) => sum + s.lectures.reduce((aSum, l) => aSum + (l.assignments?.length || 0), 0),
      0
    );
    
    console.log(`Total lectures: ${totalLectures}`);
    console.log(`Total assignments: ${totalAssignments}`);
    
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

main();

File: add_harvard_course.py
import json
import os
import psycopg2
from psycopg2.extras import execute_values
from uuid import uuid4
from datetime import datetime

# Database connection - you'll need to set these environment variables
DB_HOST = os.getenv('DB_HOST', 'localhost')
DB_PORT = os.getenv('DB_PORT', '5432')
DB_NAME = os.getenv('DB_NAME', 'postgres')
DB_USER = os.getenv('DB_USER', 'postgres')
DB_PASSWORD = os.getenv('DB_PASSWORD', '')

def load_course_data():
    """Load course data from JSON file"""
    with open('harvard_python_course.json', 'r', encoding='utf-8') as f:
        return json.load(f)

def get_db_connection():
    """Create database connection"""
    return psycopg2.connect(
        host=DB_HOST,
        port=DB_PORT,
        database=DB_NAME,
        user=DB_USER,
        password=DB_PASSWORD
    )

def insert_course(conn, course_data):
    """Insert course into video_courses table"""
    course_id = str(uuid4())
    now = datetime.now()
    
    with conn.cursor() as cur:
        cur.execute("""
            INSERT INTO public.video_courses 
            (id, title, slug, description, short_description, cover_url, category, difficulty, price_gel, is_active, sort_order, created_at, updated_at)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            ON CONFLICT (slug) DO UPDATE SET
                title = EXCLUDED.title,
                description = EXCLUDED.description,
                short_description = EXCLUDED.short_description,
                category = EXCLUDED.category,
                difficulty = EXCLUDED.difficulty,
                price_gel = EXCLUDED.price_gel,
                is_active = EXCLUDED.is_active,
                sort_order = EXCLUDED.sort_order,
                updated_at = EXCLUDED.updated_at
            RETURNING id
        """, (
            course_id,
            course_data['title'],
            course_data['slug'],
            course_data['description'],
            course_data['short_description'],
            course_data.get('cover_url'),
            course_data['category'],
            course_data['difficulty'],
            course_data['price_gel'],
            course_data['is_active'],
            course_data['sort_order'],
            now,
            now
        ))
        
        # If course already exists, get its ID
        result = cur.fetchone()
        course_id = result[0] if result else course_id
        
    conn.commit()
    return course_id

def insert_sections(conn, course_id, sections):
    """Insert sections into video_course_sections table"""
    section_ids = {}
    
    for section in sections:
        section_id = str(uuid4())
        now = datetime.now()
        
        with conn.cursor() as cur:
            cur.execute("""
                INSERT INTO public.video_course_sections 
                (id, course_id, title, sort_order, created_at)
                VALUES (%s, %s, %s, %s, %s)
                ON CONFLICT DO NOTHING
                RETURNING id
            """, (
                section_id,
                course_id,
                section['title'],
                section['sort_order'],
                now
            ))
            
            result = cur.fetchone()
            section_id = result[0] if result else section_id
            
        section_ids[section['title']] = section_id
        
    conn.commit()
    return section_ids

def insert_lectures_and_assignments(conn, course_id, section_ids, sections):
    """Insert lectures and assignments into video_lectures and video_assignments tables"""
    for section in sections:
        section_id = section_ids[section['title']]
        
        for lecture in section['lectures']:
            lecture_id = str(uuid4())
            now = datetime.now()
            
            # Insert lecture
            with conn.cursor() as cur:
                cur.execute("""
                    INSERT INTO public.video_lectures 
                    (id, section_id, course_id, title, description, video_url, video_storage_path, duration_seconds, sort_order, is_free_preview, created_at)
                    VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                    ON CONFLICT DO NOTHING
                    RETURNING id
                """, (
                    lecture_id,
                    section_id,
                    course_id,
                    lecture['title'],
                    lecture['description'],
                    lecture.get('video_url'),
                    lecture.get('video_storage_path'),
                    lecture.get('duration_seconds', 0),
                    0,  # sort_order - will be updated based on position
                    lecture.get('is_free_preview', False),
                    now
                ))
                
                result = cur.fetchone()
                lecture_id = result[0] if result else lecture_id
            
            # Insert assignments for this lecture
            if 'assignments' in lecture:
                for assignment in lecture['assignments']:
                    assignment_id = str(uuid4())
                    
                    with conn.cursor() as cur:
                        cur.execute("""
                            INSERT INTO public.video_assignments 
                            (id, lecture_id, course_id, title, description, sort_order, created_at)
                            VALUES (%s, %s, %s, %s, %s, %s, %s)
                            ON CONFLICT DO NOTHING
                        """, (
                            assignment_id,
                            lecture_id,
                            course_id,
                            assignment['title'],
                            assignment['description'],
                            0,  # sort_order
                            now
                        ))
    
    conn.commit()

def main():
    """Main function to add course to database"""
    print("Loading course data...")
    course_data = load_course_data()
    
    print("Connecting to database...")
    conn = get_db_connection()
    
    try:
        print("Inserting course...")
        course_id = insert_course(conn, course_data['course'])
        print(f"Course ID: {course_id}")
        
        print("Inserting sections...")
        section_ids = insert_sections(conn, course_id, course_data['sections'])
        print(f"Inserted {len(section_ids)} sections")
        
        print("Inserting lectures and assignments...")
        insert_lectures_and_assignments(conn, course_id, section_ids, course_data['sections'])
        
        print("Course successfully added to database!")
        print(f"Course: {course_data['course']['title']}")
        print(f"Slug: {course_data['course']['slug']}")
        print(f"Total sections: {len(course_data['sections'])}")
        
        total_lectures = sum(len(section['lectures']) for section in course_data['sections'])
        total_assignments = sum(len(lecture.get('assignments', [])) for section in course_data['sections'] for lecture in section['lectures'])
        print(f"Total lectures: {total_lectures}")
        print(f"Total assignments: {total_assignments}")
        
    except Exception as e:
        print(f"Error: {e}")
        conn.rollback()
        raise
    finally:
        conn.close()

if __name__ == "__main__":
    main()
